src/api/removeOrphanPkgs.ts

```diff
@@ -48,6 +52,26 @@ export default async function removeOrphanPkgs (
   if (!opts.dryRun) {
     if (notDependents.length) {
       await Promise.all(notDependents.map(async notDependent => {
+        if (opts.shamefullyFlatten && opts.oldShrinkwrap.packages) {
```

Comment:
I guess it is not critical but maybe the mapping could be duplicated and inserted into a `if (opts.shamefully...` to reduce impact on existing code.

```ts
 if (opts.shamefullyFlatten && opts.oldShrinkwrap.packages) {
   await Promise.all(notDependents.map(async notDependent => {
      const pkgShr = opts.oldShrinkwrap.packages[dp.relative(opts.oldShrinkwrap.registry, notDependent)]
      const pkgInfo = getPkgInfoFromShr(notDependent, pkgShr)
      const rootLink = path.join(rootModules, pkgInfo.name)
      // rootLink might not exist anymore because it might have been delete earlier
      if (await exists(rootLink)) {
        const linkTarget = path.relative(rootModules, await resolveLinkTarget(rootLink))
        // we only want to remove the root link if it points to a version that we are removing
        if (linkTarget.startsWith(`.${notDependent}`)) {
          await removeTopDependency({
            name: pkgInfo.name,
            dev: false,
            optional: false,
          }, {
            modules: rootModules,
            bin: opts.bin
          })
        }
      }
    }))
 }
```